### DISCUSSION TITLE
What's the quality of the Audio Files?

### DISCUSSION THREAD

**Discussion by amsuarez-sz** [shaked6540/YoutubePlaylistDownloader#114] [Q&A]
First of all, Congratulations! It's an amazing tool.
I'm trying to get some podcasts from music.youtube. I wonder what's the quality of the files?
I tried setting the bit rate to a higher value but I guess it is downloaded in 124 k and then converted to that specified value.
Is there a way to set the best quality available? 

Thanks in advance

**Comment by shaked6540** [comment]
The quality differs for each video, generally I'm checking all the available streams from youtube and picking the one with the highest bitrate. At the time I wrote it I wasn't sure which stream would be the best quality, and its probably more than just picking the highest bitrate (like for example, an opus stream with 128kbps would be better than an mp3 stream with 132kbps). I might rewrite that section if I find it picking low quality audio in the future, but so far it looks like its working good

**Comment by amsuarez-sz** [comment]
Thank you. I tried the Youtube Music Streaming service (not just the ordinary youtube version) which provides music (the same way Spotify/Deezer do) and the program works perfectly but somehow I can only get the 128 kbps when it is set 320 kbps on the Youtube Settings (maybe because I am logged in)

**Comment by shaked6540** [comment]
Can you screenshot your settings?

**Reply by amsuarez-sz** [reply]
This is how the settings looks from Youtube
![image](https://user-images.githubusercontent.com/80419754/217245184-7b9f4b75-2733-4cb5-9642-4ce33661b97c.png)

And this is from your program
![image](https://user-images.githubusercontent.com/80419754/217245440-ad04a017-4852-44f1-9c77-73092df5aa91.png)

I guess the program tries to convert the downloaded audio (in 125 kbps) to 320 kbps.
But it is still 125 in a higher bitrate. 
I used to get the higher audio using Loader.to but that site is not safe for many reasons
Or maybe something changed from the YT side.

I hope this helps a bit

**Reply by shaked6540** [reply]
Oh, I thought you meant that the output file is not in 320kbps

Like I said, the highest bitrate is not always available, the program just picks the highest one it can get. If you're logged in to youtube it doesn't affect the program, it can't know that you are logged in (its like opening an incognito browser and going to youtube)

I don't support logging in to youtube because of security concerns

**Reply by amsuarez-sz** [reply]
I see and maybe that's the reason. When I'm logged in, I can play the files with my settings. And maybe youtube doesn't deliver the highest resolution when you open an incognito browser. 

Just have one last question. Is there a way to hardcode the program and insert my credentials if I run it on Visual Studio?

**Reply by shaked6540** [reply]
Yep, it wouldn't be too hard I assume. You'll have to extract the credentials from YouTube when you login with the browser and create a new HttpClient, set the default cookies to your crendentials and pass it to the YoutubeClient at https://github.com/shaked6540/YoutubePlaylistDownloader/blob/master/YoutubePlaylistDownloader/GlobalConsts.cs#L74 

Use this constructor: https://github.com/Tyrrrz/YoutubeExplode/blob/master/YoutubeExplode/YoutubeClient.cs#L38

[This is a good way to add cookies to an HttpClient](https://stackoverflow.com/a/56822712/4334327) on every request, create it in the static constructor of GlobalConsts and also create a global YoutubeClient variable with it and in line 74 set it to return the YoutubeClient you created


After that, all the different modules in the program will use your credentials. But beware, downloading youtube videos might flag your account for spam and if your youtube account gets banned you will also lose your gmail account (and every other google service linked to it)

I don't know exactly which cookies youtube use for authentication, so you will either have to guess or just add them all.
